tests/CMakeLists.txt

```diff
@@ -1,14 +1,29 @@
 # Add a test for a given source file for each version of OpenCL
 function(add_header_test NAME SOURCE)
+  # Some headers include system files which require anonymous structs
+  # so we cannot test them with /Za:
+  set(TEST_SKIP_ANSI_TESTING
+    cl_d3d10_h cl_d3d11_h cl_dx9_media_sharing_h cl_dx9_media_sharing_intel_h cl_icd_h cl_layer_h)
+  get_filename_component(LANG_DIR "${CMAKE_CURRENT_SOURCE_DIR}" NAME)
+  if(LANG_DIR MATCHES cpp)
+    set(CPP_SOURCE_PATH "${CMAKE_CURRENT_BINARY_DIR}/${SOURCE}pp")
+    file(RELATIVE_PATH REL_SOURCE_PATH "${CMAKE_CURRENT_BINARY_DIR}" "${CMAKE_CURRENT_SOURCE_DIR}/../${SOURCE}")
+    file(TO_NATIVE_PATH "${REL_SOURCE_PATH}" NATIVE_SOURCE_PATH)
+    file(WRITE
+      "${CPP_SOURCE_PATH}"
+      "#include \"${NATIVE_SOURCE_PATH}\"\n"
+    )
+    set(SOURCE_PATH "${CPP_SOURCE_PATH}")
+    set(LANG cpp)
```

Comment:
Should we just add the .cpp files vs relying on cmake trickery to do this for us?

Reply:
This was the order of things I tried:
- First, just add two executables in the same script file and for one of them tweak the source file lang. Source file languages however are controlled at directory level (`CMAKE_CURRNENT_LIST_DIR` level), so that changes it for both.
- Moving the scripts to separate directories and controlling the source lang that way introduces a warning on Clang saying [`warning: treating ‘A’ input as ‘B’ when in C++ mode, this behavior is deprecated`](https://clang.llvm.org/docs/DiagnosticsReference.html#wdeprecated) which can only be silenced bundled with other warnings and not selectively.
- Didn't know which was the better approach, to stamp out single-line files and double the number of files in the repo or generate them via CMake. Both are simple conceptually. One looks ugly in code, the other looks ugly on the command-line/File Explorer/Solution Explorer.